ublksrv_tgt.cpp

```diff
@@ -756,6 +770,113 @@ static void cmd_dev_list_usage(char *cmd)
 	printf("%s list [-n DEV_ID]\n", cmd);
 }
 
+static int __cmd_dev_user_recover(int number, bool verbose)
+{
+	struct ublksrv_dev_data data = {
+		.dev_id = number,
+		.run_dir = UBLKSRV_PID_DIR,
+	};
+	struct ublksrv_ctrl_dev *dev;
+	struct ublksrv_tgt_base_json tgt_json = {0};
+	const char *buf;
+	char pid_file[64];
+	int ret;
+
+	dev = ublksrv_ctrl_init(&data);
+
+	ret = ublksrv_ctrl_start_recovery(dev);
+	if (ret < 0) {
+			fprintf(stderr, "can't start recovery for %d\n", number);
+		goto fail;
+	}
+
+	buf = ublksrv_tgt_get_dev_data(dev);
+	if (!buf) {
+		fprintf(stderr, "get dev %d data failed\n", number);
+		ret = -1;
+		goto fail;
+	}
+
+	ret = ublksrv_json_read_dev_info(buf, &dev->dev_info);
+	if (ret < 0) {
+		fprintf(stderr, "can't read dev info for %d\n", number);
+		goto fail;
+	}
+	ret = ublksrv_json_read_target_base_info(buf, &tgt_json);
+	if (ret < 0) {
+		fprintf(stderr, "can't read dev info for %d\n", number);
+		goto fail;
+	}
+
+	snprintf(pid_file, 64, "%s/%d.pid", dev->run_dir, dev->dev_info.dev_id);
+	ret = unlink(pid_file);
+	if (ret < 0) {
+		fprintf(stderr, "can't delete old pid_file for %d, error:%s\n",
+				number, strerror(errno));
+		goto fail;
+	}
+
+	/*
+	 * TODO: read target specific info such as "backing_file"
```

Comment:
Here the whole target info should be recovered, otherwise the target can't be in workable state.

Reply:
I konw this problem. Maybe we should add null_reinit_tgt() and loop_reinit_tgt() to read target-sprcific info such as "backing_file" while creating new daemon process.

Reply:
The problem should not be hard to solve, mark current state as recovering, and pass the json buffer to target code via 'struct ublksrv_ctrl_dev'. The API isn't locked yet, so we still can add field
now. But I plan to lock it in 1.0 release.

The recover code in *_init_tgt() is pretty easy actually, just open the backing file, set correct file size, then copy the json buffer.